flink-core/src/main/java/org/apache/flink/core/classloading/MutableURLClassLoader.java

```diff
@@ -0,0 +1,152 @@
+/*
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.apache.flink.core.classloading;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.CoreOptions;
+import org.apache.flink.util.FlinkUserCodeClassLoader;
+import org.apache.flink.util.FlinkUserCodeClassLoaders;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.apache.flink.util.FlinkUserCodeClassLoader.NOOP_EXCEPTION_HANDLER;
+
+/**
+ * This classloader delegates to the actual user classloader {@link FlinkUserCodeClassLoader}. Upon
+ * the existed function of {@link FlinkUserCodeClassLoader}, it also provides extra {@link
+ * #addURL(URL)} and {@link #removeURL(URL)} methods, this allows user to add and remove jar
+ * dynamically.
+ *
+ * <p>User can specify the class load order by set the parameter (classloader.resolve-order) in
+ * {@link Configuration}, default order is child-first.
+ */
+public class MutableURLClassLoader extends URLClassLoader {
```

Comment:
When will users directly create this classloader?

Reply:
Currently, only table module need it such as TableEnvironment.

Reply:
I was asking this question to see if if we should mark it a `@Internal` or `@PublicEnvolving`. If the class is only seen and used by Flink internally, we should mark it as `@Internal`.